File: spannablehelper/src/main/java/com/dc/spannablehelper/span/CenterVerticalImageSpan.java
package com.dc.spannablehelper.span;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.text.style.ImageSpan;

import org.jetbrains.annotations.NotNull;

/**
 * Create by CZH
 * Create date 2020/10/10
 * Description:居中显示的imageSpan
 */
public class CenterVerticalImageSpan extends ImageSpan {
    private float marginLeft = 0;
    private float marginRight = 0;

    public CenterVerticalImageSpan(Drawable drawable, int verticalAlignment, float marginLeft, float marginRight) {
        super(drawable, verticalAlignment);
        this.marginLeft = marginLeft;
        this.marginRight = marginRight;
    }



    /**
     * update the text line height
     */
    @Override
    public int getSize(@NotNull Paint paint, CharSequence text, int start, int end,
                       Paint.FontMetricsInt fontMetricsInt) {
        Drawable drawable = getDrawable();
        Rect rect = drawable.getBounds();
        if (fontMetricsInt != null) {
            Paint.FontMetricsInt fmPaint = paint.getFontMetricsInt();
            int fontHeight = fmPaint.descent - fmPaint.ascent;
            int drHeight = rect.bottom - rect.top;
            int centerY = fmPaint.ascent + fontHeight / 2;

            fontMetricsInt.ascent = centerY - drHeight / 2;
            fontMetricsInt.top = fontMetricsInt.ascent;
            fontMetricsInt.bottom = centerY + drHeight / 2;
            fontMetricsInt.descent = fontMetricsInt.bottom;
        }
        return (int) (rect.right + marginLeft + marginRight);
    }


    @Override
    public void draw(Canvas canvas, CharSequence text, int start, int end,
                     float x, int top, int y, int bottom, Paint paint) {
        Drawable drawable = getDrawable();
        canvas.save();
        Paint.FontMetricsInt fmPaint = paint.getFontMetricsInt();
        int fontHeight = fmPaint.descent - fmPaint.ascent;
        int centerY = y + fmPaint.descent - fontHeight / 2;
        int transY = centerY - (drawable.getBounds().bottom - drawable.getBounds().top) / 2;
        canvas.translate(x + marginLeft, transY);
        drawable.draw(canvas);
        canvas.restore();
    }
}

File: spannablehelper/src/main/java/com/dc/spannablehelper/span/CenterVerticalImageSpan.kt
package com.dc.spannablehelper.span

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Paint.FontMetricsInt
import android.graphics.drawable.Drawable
import android.text.style.ImageSpan

/**
 * Create by CZH
 * Create date 2020/10/10
 * Description:居中显示的imageSpan
 */
class CenterVerticalImageSpan(drawable: Drawable, verticalAlignment: Int, private val marginLeft: Float,
                              private val marginRight: Float) : ImageSpan(drawable, verticalAlignment) {


    /**
     * update the text line height
     */
    override fun getSize(paint: Paint, text: CharSequence, start: Int, end: Int, fontMetricsInt: FontMetricsInt?): Int {
        val drawable = drawable
        val rect = drawable.bounds
        if (fontMetricsInt != null) {
            val fmPaint = paint.fontMetricsInt
            val fontHeight = fmPaint.descent - fmPaint.ascent
            val drHeight = rect.bottom - rect.top
            val centerY = fmPaint.ascent + fontHeight / 2
            fontMetricsInt.ascent = centerY - drHeight / 2
            fontMetricsInt.top = fontMetricsInt.ascent
            fontMetricsInt.bottom = centerY + drHeight / 2
            fontMetricsInt.descent = fontMetricsInt.bottom
        }
        return (rect.right + marginLeft + marginRight).toInt()
    }

    override fun draw(canvas: Canvas, text: CharSequence, start: Int, end: Int, x: Float, top: Int, y: Int, bottom: Int,
                      paint: Paint) {
        val drawable = drawable
        canvas.save()
        val fmPaint = paint.fontMetricsInt
        val fontHeight = fmPaint.descent - fmPaint.ascent
        val centerY = y + fmPaint.descent - fontHeight / 2
        val transY = centerY - (drawable.bounds.bottom - drawable.bounds.top) / 2
        canvas.translate(x + marginLeft, transY.toFloat())
        drawable.draw(canvas)
        canvas.restore()
    }
}